lib/Dialect/TTMetal/Transforms/Passes.cpp

```diff
@@ -272,34 +272,42 @@ class TTIRToTTMetalLayoutRewriter : public OpRewritePattern<ttir::ToLayoutOp> {
     tensixBlock->addArgument(inputCBTy, op.getLoc());
     tensixBlock->addArgument(outputCBTy, op.getLoc());
 
-    int shardTileVolume = 1;
-    for (auto dim : (shouldTilize ? outputLayout.getMemref().getShape()
-                                  : inputLayout.getMemref().getShape())) {
-      shardTileVolume *= dim;
-    }
+    llvm::ArrayRef<int64_t> shardTileShape =
+        (shouldTilize ? outputLayout.getMemref().getShape()
+                      : inputLayout.getMemref().getShape());
+
+    assert(shardTileShape.size() == 2 && "Tile shape rank must be 2");
 
-    auto numTiles = tensixBuilder.create<arith::ConstantOp>(
-        op.getLoc(), tensixBuilder.getI32Type(),
-        tensixBuilder.getI32IntegerAttr(shardTileVolume));
+    // How many tiles should kernel tilize in one block.
+    arith::ConstantOp numTilesPerBlock =
+        tensixBuilder.create<arith::ConstantOp>(
+            op.getLoc(), tensixBuilder.getI32Type(),
+            tensixBuilder.getI32IntegerAttr(shardTileShape[1]));
```

Comment:
This should be `shardTileShape[shardTileShape.size() - 1]` per rank N comment above.